### PR TITLE
Added bandwidth selection of Andrews (1991) and of Newey - West (1994)

### PR DESCRIPTION
**Pull request by Mill7**

### CODE REVIEW

File: arch/unitroot/unitroot.py [hunk 1]
```diff
@@ -1714,3 +1714,63 @@ def kpss_crit(stat, trend='c'):
     crit_value = interp(cv, y[::-1], x[::-1])
 
     return pvalue, crit_value
+
+  def BandWidthNW(y, kernel = "ba"):
```
Comment (bashtage) [inline arch/unitroot/unitroot.py]: Should be `autobandwitdh` or `auto_bandwidth` (probablye the latter).

So space around = for default args.

File: arch/unitroot/unitroot.py [hunk 2]
```diff
@@ -1714,3 +1714,63 @@ def kpss_crit(stat, trend='c'):
...
+    '''
+    | Automatic bandwidth selection of Andrews (1991) and of Newey and West (1994)
```
Comment (bashtage) [inline arch/unitroot/unitroot.py]: No `|` in doc string.  The docstring must follow the NumPy style

https://numpydoc.readthedocs.io/en/latest/format.html

with sections 

```
Parameters
----------

Returns
-------

Notes
-----
```
etc.

File: arch/unitroot/unitroot.py [hunk 3]
```diff
@@ -1714,3 +1714,63 @@ def kpss_crit(stat, trend='c'):
...
+    |  - "qs": Quadratic Spectral kernel
+    | 
+    | return the Bandwidth
+    '''
+
+    n = int(4 * np.power(len(y)/100, 2/9))       
+    sig = (n+1)*[0]
+
+    for i in range(n+1):
+
```
Comment (bashtage) [inline arch/unitroot/unitroot.py]: I prefer tigheer spacing.  Only use empty lines sparingly to separate logical blocks of code, not with, e.g., a loop

File: arch/unitroot/unitroot.py [hunk 4]
```diff
@@ -1714,3 +1714,63 @@ def kpss_crit(stat, trend='c'):
...
+        sig[i] = np.sum(np.dot(y[i:], y[:len(y) - i]) )
+
+    sigmaM1 = sig[1:len(sig)] #sigma without the 1st element
```
Comment (bashtage) [inline arch/unitroot/unitroot.py]: Two spaces before in-line comment, and a space after the #

File: arch/unitroot/unitroot.py [hunk 5]
```diff
@@ -1714,3 +1714,63 @@ def kpss_crit(stat, trend='c'):
...
+
+    n = int(4 * np.power(len(y)/100, 2/9))       
+    sig = (n+1)*[0]
+
+    for i in range(n+1):
+
+        sig[i] = np.sum(np.dot(y[i:], y[:len(y) - i]) )
+
+    sigmaM1 = sig[1:len(sig)] #sigma without the 1st element    
+    s0 = sig[0] + 2 * np.sum(sigmaM1)
```
Comment (bashtage) [inline arch/unitroot/unitroot.py]: Pythonic names, so sigmaM1 -> sigma_m1

File: arch/unitroot/unitroot.py [hunk 6]
```diff
@@ -1714,3 +1714,63 @@ def kpss_crit(stat, trend='c'):
...
+        s2 = 0
+
+        for j in range(len(sigmaM1)):
+            s2 += np.power((j+1), 2)*sigmaM1[j] 
+
+        s2 *= 2
+        q = 2  
+
+
+    Tpower = 1 / (2 * q + 1)
```
Comment (bashtage) [inline arch/unitroot/unitroot.py]: Tpower -> t_power

File: arch/unitroot/unitroot.py [hunk 7]
```diff
@@ -1714,3 +1714,63 @@ def kpss_crit(stat, trend='c'):
...
+    Tpower = 1 / (2 * q + 1)    
+
+    if kernel == "ba":        
+        gamma = 1.1447 * np.power((np.power(s1 / s0, 2)),Tpower)            
+    elif kernel == "pa" :        
+         gamma = 2.6614 * np.power((np.power(s2 / s0, 2)),Tpower)         
+    elif kernel == "qs":        
+         gamma = 1.3221 * np.power((np.power(s2 / s0, 2)),Tpower)   
+
+    bandwidth = gamma * np.power(len(y), Tpower)
```
Comment (bashtage) [inline arch/unitroot/unitroot.py]: Aren't these all scalars?  Can you just use `**`?

File: arch/unitroot/unitroot.py [hunk 8]
```diff
@@ -1714,3 +1714,63 @@ def kpss_crit(stat, trend='c'):
...
+    | 
+    | y: Data on which to apply the bandwidth selction
+    | kernel: The kernel function to use for selecting the bandwidth
+    |  - "ba": Bartlett kernel (default value)
+    |  - "pa": Parzen kernel
+    |  - "qs": Quadratic Spectral kernel
+    | 
+    | return the Bandwidth
+    '''
+
```
Comment (bashtage) [inline arch/unitroot/unitroot.py]: need a call to `y = np.asarray(y)` here to strip and pandas data sets.  Also need to check that `y.ndim==1` and `y` has more than `n` observations.

**Comment by bashtage** [comment]
Take a look at the [PEP8 speaks post above](https://github.com/bashtage/arch/pull/303#issuecomment-508499573).  These all need to be sorted too.

File: arch/unitroot/unitroot.py (hunk 8 above)
Comment (Mill7) [reply arch/unitroot/unitroot.py]: Now Raise an Exception when y has not the expected properties

**Comment by bashtage** [comment]
@Mill7 This is dueo tlinting errors:

arch/unitroot/unitroot.py:1770:23: W291 trailing whitespace
arch/unitroot/unitroot.py:1771:37: E225 missing whitespace around operator

Those are the line numbers.  You need to remove any white space at the end of a line, and add a space around the operator on 1771.

File: arch/unitroot/unitroot.py [hunk 9]
```diff
@@ -1714,3 +1714,68 @@ def kpss_crit(stat, trend='c'):
...
+        q = 1
+    elif kernel in ["pa", "qs"]:
+        s2 = 0
+        for j in range(len(sigma_m1)):
+            s2 += ((j+1)**2)*sigma_m1[j]
+        s2 *= 2
+        q = 2
+    t_power = 1/(2*q+1)
+
+    if kernel == "ba":
```
Comment (bashtage) [inline arch/unitroot/unitroot.py]: Seem to have a space after the : here.

File: arch/unitroot/unitroot.py [hunk 10]
```diff
@@ -1714,3 +1714,68 @@ def kpss_crit(stat, trend='c'):
...
+    elif kernel in ["pa", "qs"]:
+        s2 = 0
+        for j in range(len(sigma_m1)):
+            s2 += ((j+1)**2)*sigma_m1[j]
+        s2 *= 2
+        q = 2
+    t_power = 1/(2*q+1)
+
+    if kernel == "ba":       
+        gamma = 1.1447 * (((s1 / s0)** 2)**t_power)
```
Comment (bashtage) [inline arch/unitroot/unitroot.py]: Missing space around **

**Comment by bashtage** [comment]
Thanks, it is in!  The failure are related to coverage and not essential.

**Pull request merged**